packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  // If there is no ticket creation date or no actions, we return an empty timeline.
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const sortedActions: rawUserAction[] = sortActions(actions);
+
+  const mappedUserActions = mapUserActions(sortedActions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedUserActions,
+    ticketCreationDate
+  );
+
+  // Filter out custom actions to be excluded from the timeline
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+// Sorting actions by most recent
+export const sortActions = (actions: rawUserAction[]) => {
+  const sortedActionsByMostRecent = (a: rawUserAction, b: rawUserAction) =>
+    Number(b.time) - Number(a.time);
+
+  return [...actions].sort(sortedActionsByMostRecent);
+};
+
+export const filterActions = (
+  actions: UserAction[],
+  excludedCustomActions: string[]
+): UserAction[] => {
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
```

Comment:
If we are excluding only custom events, how to exclude empty searches for example?

Reply:
The current logic was to filter out custom events or events by specific `eventValue` & `eventType` and not for searches. So if there are more events we want to filter out we should sync and decide which ones can be excludable

Reply:
Yes we discussed it in the live review of your pr and you need to filter out the empty searches too.

Reply:
As discussed above too I would make a method that deals with a single action instead of a method that deals with ALL actions.

```
export const shouldKeepAction = (action: UserAction, excludedCustomActions: string[]): boolean => !excludedCustomActions.includes(action.type);
```
So you can then call it like this

```
const filteredActions = actions.filter((action) => shouldKeepAction(action, excludedCustomActions));
```